src/CacheClient.cpp

```diff
@@ -0,0 +1,145 @@
+
+static int SlowCallback(void *user_data, const char* label)
+{
+    SlowCallbackData *data = (SlowCallbackData *)user_data;
+    MutexLock(data->queue_lock);
+    char buffer[1000];
+    snprintf(buffer,sizeof(buffer),"%s %s", data->node_data->m_Annotation.Get(), label);
+    int sendNextCallbackIn = PrintNodeInProgress(data->node_data, data->time_of_start, data->build_queue, buffer);
+    MutexUnlock(data->queue_lock);
+    return sendNextCallbackIn;
+}
+
+static int SlowCallback_CacheRead(void *user_data)
+{
+    return SlowCallback(user_data, "[CacheRead]");
+}
+
+static int SlowCallback_CacheWrite(void *user_data)
+{
+    return SlowCallback(user_data, "[CacheWrite]");
+}
+
+enum Operation
+{
+    kOperationRead,
+    kOperationWrite
+};
+
+static bool Invoke_REAPI_Cache_Client(const HashDigest& digest, StatCache *stat_cache, const FrozenArray<FrozenFileAndHash>& outputFiles, ThreadState* thread_state, Operation operation, const Frozen::DagNode* dagNode, Mutex* queue_lock)
+{
+    ProfilerScope profiler_scope("InvokeCacheMe", thread_state->m_ProfilerThreadId, outputFiles[0].m_Filename);
+
+    const char* reapi_raw = getenv(kENV_REAPI_CACHE_CLIENT);
+    if (reapi_raw == nullptr)
+        return false;
+
+    PathBuffer pathbuf;
+    PathInit(&pathbuf, reapi_raw);
+    char reapi[kMaxPathLength];
+    PathFormat(reapi, &pathbuf);
+
+    char buffer[5000];
+    int totalWritten = 0;
+    char digestString[kDigestStringSize];
+    DigestToString(digestString, digest);
+
+    const char* cmd = operation == kOperationRead ? "down" : "up";
+    totalWritten += snprintf(buffer, sizeof(buffer), "%s %s %s00000000000000000000000000000002", reapi, cmd, digestString);
+
+    //when we start caching nodes with tons of outputs, we should move the filelist to a separate file. for now this will do,
+    for (auto &it : outputFiles)
+    {
+        PathBuffer output;
+        PathInit(&output, it.m_Filename);
+        MakeDirectoriesForFile(stat_cache, output);
+        int remainingBudget = sizeof(buffer)-totalWritten;
+
+        //headsup: requiredSpace return value does _not_ include length of the 0 terminator.
+        int requiredSpace = snprintf(buffer+totalWritten, remainingBudget, " \"%s\" ", it.m_Filename.Get());
+
+        if (requiredSpace >= remainingBudget)
+        {
+            Log(kError, "Building CacheClient string exceeded buffer length");
+            return false;
+        }
+
+        totalWritten += requiredSpace;
+    }
+
+    SlowCallbackData slowCallbackData;
+    slowCallbackData.node_data = dagNode;
+    slowCallbackData.time_of_start = TimerGet();
+    slowCallbackData.queue_lock = queue_lock;
+    slowCallbackData.build_queue = thread_state->m_Queue;
+
+    Log(kDebug,"%s\n",buffer);
+    ExecResult result = ExecuteProcess(buffer, 0, nullptr, nullptr, thread_state->m_ThreadIndex, false, operation == kOperationRead ? SlowCallback_CacheRead : SlowCallback_CacheWrite , &slowCallbackData);
```

Comment:
Also, we should pass `&thread_state->m_LocalHeap` for the forth argument - or we will fail on the mac with this error:
```        CroakAbort("Either pass in a heap so we can allocate buffers to store stdout, or ask to stream directly to stdout");
```